### PR TITLE
Fix linking issue when logging is enabled

### PR DESCRIPTION
**Pull request by elecro**
Make the _log method implementation into a template method
to avoid linking issues. Also remove duplicate defines.

Fixes: #61

Signed-off-by: Peter Gal <galpeter@inf.u-szeged.hu>

### CODE REVIEW

File: src/utils/gepard-defs.h [hunk 1]
```diff
@@ -41,77 +43,51 @@
...
-} while (false)
-
-void _log(int level, std::string msg) {
+template<typename T>
+void _log(unsigned int level, T& msg) {
   const int max_level = LOG_LEVEL;
   if (level <= max_level) {
-    std::cout << msg << std::endl;
+    static char *color[] = { LOG1_COLOR, LOG2_COLOR, LOG3_COLOR };
+    std::cout << color[level] << msg << CLEAR_COLOR << std::endl;
```
Comment (szledan) [inline src/utils/gepard-defs.h]: The problem this way is that, it doesn't support a complex logging:
```c++
int x = 1;
void* ptr = Class::create();
Float value = ptr->fx(x) * x;
LOG1("The value is ptr[" << ptr << "].fx(x=" << x << ") * x = " << value << ".");
```

Comment (szledan) [inline src/utils/gepard-defs.h]: But at the moment your patch is solved this issue. We need to discuss which logging system should we use?

Comment (elecro) [inline src/utils/gepard-defs.h]: Well it is possible to add back an intermediate code part which converts everything to string if that is desired.

**Comment by elecro** [comment]
@szledan I've done a bit of update PTAL.

**Comment by szledan** [comment]
LGTM (all tests and builds pass locally)

**Comment by dbatyai** [comment]
LGTM

**Pull request merged**